### PR TITLE
Fixes Coordinated Omission

### PR DESCRIPTION
**Pull request by franz1981**

### CODE REVIEW

**Comment by franz1981** [comment]
Need some code review here: @shawkins, I didn't yet tried on real machines to see the impact (+ if it works :P)

**Comment by shawkins** [comment]
> Need some code review here: @shawkins, I didn't yet tried on real machines to see the impact (+ if it works :P)

It seems fine.  But unless I'm missing something this seems to only be measuring the delay introduced by rate limiting.  Is the intent here to introduce better logic for determining the maximum sustainable production rate?

**Comment by shawkins** [comment]
@franz1981 I've updated the repo to confluent's latest - there really wasn't much in those 12 commits.  

On this pr I may still be missing something, but the more I think about it seems like this would only help you identify when producers cannot deliver the amount of load specified.  That is if you request a rate of X messages per second, but only see Y messages per second when the producer delay is close to 0, then you would know that the producers aren't able of producing that load.  Values above zero would help you identify if you have more producers than needed for the given load, but I don't see that it would help identify the maximum sustainable rate.

On the fork we are measuring the producer average queue time and throttle times via Kafka metrics, so that we have those values directly - it's the average queue time that would help determine the max sustainable production rate.  The max sustainable production rate is when the average queue time is within some tolerance of the linger - if it's below the linger, that should indicate you are filling batches more quickly than expected.  If it's above the linger, then items are waiting in the queue for longer than needed.

**Comment by franz1981** [comment]
Yep, in the current form this is not meant to help to find the sustainable rate, but can be done in a follow up pr (I would like this to be first merged upstream).

The main benefit in the current form is to spot weird single pauses of clients/os/machine/jvm (on client it can happen to retrieve metadata or due to existed buffer pools) that can affect SLA. Metrics pretty valuable for low latency benchmark runs IMO

**Comment by shawkins** [comment]
> In short: high delay means that the other latency metrics cannot be trusted

How can you tell a difference between expected and unexpected publisher delay?  Even if the value is high, that would indicate an unexpected pause with rate limiting - how could you make conclusions about the other latencies based upon that?.

**Comment by franz1981** [comment]
> How can you tell a difference between expected and unexpected publisher delay?

The change is replacing the existing google rate limiter, that's unaware of the scheduling of requests with an ad-hoc implemented one that keep track of it, making possible to extract a so-called "intended publish time" or the time defined *by scheduling* that the load generator was supposed to initiate a publish request.
This information, together with the "real publish time" (ie the time by which the request is fired for real) allow to keep track of the publisher delays in a statistical relevant way.

**Note on statistical relevant way** 
(it's better explained on http://psy-lob-saw.blogspot.com/2016/07/fixing-co-in-cstress.html and other articles)

A single GC pause on the broker (or the client) will end up as a single outlier in the many samples collected during a run, probably appearing in the max or 99.9999XXX percentiles, and if such pause is not long enough it wouldn't impact the avg rate perceived by the load generator either (-> these are averages, hence depends by the sampling period!)...so, how to make such pauses relevant?
There are 2 ways:
- using a separate measure to keep track of *all* missed chances to keep up with the expected rate
- using such "publisher delay" to artifically enlarge both publisher and end to end latencies

For the OMB pr I've decided for the former, because way less impactfull, but other load generators (Cassandra, Yahoo YCSB, Gaitling, Wrk2, Hyperfoil) are using the latter, using the queuing theory terminology of 
```
response time = service time + wait time
```
where the response time is the artificially enlarged publisher latency (or end to end one), service time is the measured one when the fire time happens for real and wait time is the publisher delay (I've defined separately).

>how could you make conclusions about the other latencies based upon that?.

This is the point of this change: without a publisher delay info you cannot drive any conclusion related how good the load generator has been able to keep up with the expected load (averages apart, that are not precise by definition) and you cannot consider reliable the others latencies as well 
eg if the load generator has paused (because of GC pauses), coordinating with the broker, the end 2 end pauses can still be very good overall (depending when the pause happened) and on higher percentiles as well, given that, as said, a single bad outlier rarely affect a whole collection of samples.

**Comment by shawkins** [comment]
> This information, together with the "real publish time" (ie the time by which the request is fired for real) allow to keep track of the publisher delays in a statistical relevant way.

Thank you for indulging my questions.  Sorry for some confusion as a couple of my earlier comments were based upon thinking that you were looking at the time spent in the delay.

So we have:

T1 - intended send time
sleep until T1
T2 - send time

The expectation here is that T2 - T1 will be close to the minimum sleep resolution.   

What I'm still wondering is why only in this time window?  Couldn't a gc pause affect the LocalWorker executor thread not in that wait - during the producer iteration, after calls to sendAsync and before obtaining the send time?  

> This is the point of this change: without a publisher delay info you cannot drive any conclusion related how good the load generator has been able to keep up with the expected load (averages apart, that are not precise by definition) and you cannot consider reliable the others latencies as well -> they appear better then the actual "user experience"

Let me rephrase - if you can detect additional pauses, those could have have been related to or just as easily occurred during some part of a sendAsync call and therefore should be accounted for in some way.

As for how these compare to a user experience that seems hard to extrapolate - there are a lot of design choices in the test harness that likely won't line up to a user application.  The amount of parallelism in particular, but also the amount of other work being performed that influences garbage collection, etc.

On the parallelism - with the upstream version they are using Runtime.getRuntime().availableProcessors() to determine the concurrency of the producer requests.  Especially with the java 8 based image on kubernetes that's not a great choice, so in our fork we simply parallelize to the full number of producers.  That's fine as we are only running scenarios where the could be at most fewer than 100 producers per omb worker.  Obviously we should revisit that for a better cap as there was also an update to use Java 11 in the image as well.

I've also updated the pr to merge the upstream with your latest changes committed, so we'll be able to close this pr.

**Comment by franz1981** [comment]
> Thank you for indulging my questions

It's a pleasure: everytime I got to recall how it works help me to understand it a bit more :) it's a win-win

> Couldn't a gc pause affect the LocalWorker executor thread not in that wait - during the producer iteration, after calls to sendAsync and before obtaining the send time?

It's indeed a sensible question: yes it can
and that's where the notion of "schedule" come in handy; if the pause (un-noticed yet in that same producer "tick") is such big that prevent any subsequent scheduled attempt to fire "in time" the delay will be propagated to such samples.
It's exactly how a queue is supposed to act in queuing theory: queues are used to improve utilization and keep a system busy, sacrificing a bit of latency to get better throughput (and utilization); the producer client can be seen as our queue and if it delay a "bit" everything, adding some latency, is still fine: it's just important the "wait time" (time before starting to get served) won't affect the requested SLAs (ie be stick to the expected rate) where the "user experience" (mentioned before) is the "perceived latency" (including the producer delay) while delivering the expected service rate.

Will answer the rest of the questions in a separate comment :P

**Comment by shawkins** [comment]
Got everything up and running after the merge.  

The median and average  were initially fairly close to 77 and 79 ms respectively.  There were a few pauses detected in the max, on the order of 3 ms.  However after a few tests had run, the numbers got a little strange:

```
...
2022-04-08 01:01:43 [32mINFO [m [WorkloadGenerator:-1] connection.count 126.0 | Consumer Latency (ms) avg:  3.4
2022-04-08 01:01:43 [32mINFO [m [WorkloadGenerator:-1] Avg Producer Throttle Time (ms) 0.0 | Avg Producer Record Queue Time (ms) 2781.306308670891
2022-04-08 01:01:43 [32mINFO [m [WorkloadGenerator:-1] E2E Latency (ms) avg: 5499.8 - 50%: 49.0 - 99%: 22234.1 - 99.9%: 22917.1 - Max: 23057.0
2022-04-08 01:01:53 [32mINFO [m [WorkloadGenerator:-1] Pub rate 98048.2 msg/s / 95.8 MB/s | Cons rate 98076.4 msg/s / 95.8 MB/s | Backlog:  1.6 K | Pub Latency (ms) avg: 5444.8 - 50%:  7.1 - 99%: 22685.3 - 99.9%: 23799.6 - Max: 23990.8 | Pub Delay Latency (us) avg: 66767977.7 - 50%: 66781695.0 - 99%: 67102207.0 - 99.9%: 67108607.0 - Max: 67108607.0
2022-04-08 01:01:53 [32mINFO [m [WorkloadGenerator:-1] connection.count 126.0 | Consumer Latency (ms) avg:  3.5
2022-04-08 01:01:53 [32mINFO [m [WorkloadGenerator:-1] Avg Producer Throttle Time (ms) 0.0 | Avg Producer Record Queue Time (ms) 2768.3036337295507
2022-04-08 01:01:53 [32mINFO [m [WorkloadGenerator:-1] E2E Latency (ms) avg: 5463.7 - 50%: 32.0 - 99%: 22700.0 - 99.9%: 23814.0 - Max: 24018.0
2022-04-08 01:02:03 [32mINFO [m [WorkloadGenerator:-1] Pub rate 94073.3 msg/s / 91.9 MB/s | Cons rate 94147.0 msg/s / 91.9 MB/s | Backlog:  0.9 K | Pub Latency (ms) avg: 5471.2 - 50%:  6.9 - 99%: 23488.8 - 99.9%: 24638.1 - Max: 24926.0 | Pub Delay Latency (us) avg:  0.0 - 50%:  0.0 - 99%:  0.0 - 99.9%:  0.0 - Max:  0.0
...
```

I'm not sure yet what was going on there.

**Comment by franz1981** [comment]
> I'm not sure yet what was going on there

Let's have a call tomorrow; it's probably some bug of mine; as said I didn't had the chance to try it 

The easier thing I can think about is this
https://github.com/openmessaging/benchmark/pull/248#issuecomment-1055504200

eg during the sustainable rate research some publisher was left asleep while others have moved on with the new suggested (increased) rate and when they wake up the turn to have.a huge delay they will record. 

The other thing I can think about is that the suggested rate here is just too high and there's an accumulation of delay debt becuse the publishers cannot keep up.
What's the suggested rate? Is it "stable"?

Third option: I forgot some part to reset the accumulated latencies

**Comment by shawkins** [comment]
> Let's have a call tomorrow; it's probably some bug of mine; as said I didn't had the chance to try it

Thanks for catching up on this.  The takeaway is that the throughput testing that we're doing sets the production rate at an artificially high level (it would be great to add an option to simply disable throttling).  In this scenario after approximately 1 minute the values being fed into publishDelayLatencyRecorder will go out of bounds - we see increasingly large values for the delay until this happens.  After this condition is reached no further values can be added, so at the next interval 0 will be reported.  Since the exception is happening in the sendAsync handle, we are not catching that and reporting it as a test error - but it should be in the omb worker log.

Since the stat is trying to capture cumulative delay, not relative, running behind the target production at any point by 1 minute will trigger the stat collection error.

I'll go ahead and mark this pr as closed, and we can address disabling throttling or increasing the histogram value range in a follow on task.
